packages/arb-token-bridge-ui/src/hooks/useNetworks.ts

```diff
@@ -0,0 +1,212 @@
+    // E2E
+    [ChainId.Local]: local,
+    [ChainId.ArbitrumLocal]: arbitrumLocal
+  }[chainId]
+
+  return chain
+}
+
+function getPartnerChainsQueryParams(chainId: ChainId): ChainId[] {
+  try {
+    const partnerChains = getPartnerChainsForChainId(chainId)
+    return partnerChains.map(chain => chain.id)
+  } catch (e) {
+    return []
+  }
+}
+
+const getProviderForChainCache: {
+  [chainId: number]: StaticJsonRpcProvider
+} = {
+  // start with empty cache
+}
+
+function createProviderWithCache(chainId: ChainId) {
+  const rpcUrl = rpcURLs[chainId]
+  const provider = new StaticJsonRpcProvider(rpcUrl, chainId)
+  getProviderForChainCache[chainId] = provider
+  return provider
+}
+
+function getProviderForChainId(chainId: ChainId): StaticJsonRpcProvider {
+  const cachedProvider = getProviderForChainCache[chainId]
+
+  if (typeof cachedProvider !== 'undefined') {
+    return cachedProvider
+  }
+
+  return createProviderWithCache(chainId)
+}
+
+export function sanitizeQueryParams({
+  sourceChainId,
+  destinationChainId
+}: {
+  sourceChainId: ChainId | undefined
+  destinationChainId: ChainId | undefined
+}): {
+  sourceChainId: ChainId
+  destinationChainId: ChainId
+} {
+  // when both `sourceChain` and `destinationChain` are undefined, default to Ethereum and Arbitrum One
+  if (
+    typeof sourceChainId === 'undefined' &&
+    typeof destinationChainId === 'undefined'
+  ) {
+    return {
+      sourceChainId: ChainId.Ethereum,
+      destinationChainId: ChainId.ArbitrumOne
+    }
+  }
+
+  // only `destinationChainId` is defined
+  if (
+    typeof sourceChainId === 'undefined' &&
+    typeof destinationChainId !== 'undefined'
+  ) {
+    const [defaultSourceChainId] =
+      getPartnerChainsQueryParams(destinationChainId)
+    return { sourceChainId: defaultSourceChainId!, destinationChainId }
+  }
+
+  // only `sourceChainId` is defined
+  if (
+    typeof sourceChainId !== 'undefined' &&
+    typeof destinationChainId === 'undefined'
+  ) {
+    const [defaultDestinationChainId] =
+      getPartnerChainsQueryParams(sourceChainId)
+    return {
+      sourceChainId: sourceChainId,
+      destinationChainId: defaultDestinationChainId!
+    }
+  }
+
+  if (
+    !getPartnerChainsQueryParams(sourceChainId!).includes(destinationChainId!)
+  ) {
+    const [defaultDestinationChainId] = getPartnerChainsQueryParams(
+      sourceChainId!
+    )
+    return {
+      sourceChainId: sourceChainId!,
+      destinationChainId: defaultDestinationChainId!
+    }
+  }
+
+  return {
+    sourceChainId: sourceChainId!,
+    destinationChainId: destinationChainId!
+  }
+}
+
+export type UseNetworksState = {
+  sourceChain: Chain
+  sourceChainProvider: StaticJsonRpcProvider
+  destinationChain: Chain
+  destinationChainProvider: StaticJsonRpcProvider
+}
+
+export type UseNetworksSetStateParams =
+  | {
+      sourceChain: ChainId
+      destinationChain?: ChainId
+    }
+  | {
+      sourceChain?: ChainId
+      destinationChain: ChainId
+    }
```

Comment:
These should probably be `ChainId | number` since it's more accurate for custom chains